src/Discord/Internal/Types/ScheduledEvents.hs

```diff
@@ -0,0 +1,586 @@
+
+-- | Data required to issue a Modify Scheduled Event request
+-- This isnt fully type-safe, and can allow for boggus requests but I don't
+-- know of any sane solution to this
+data ModifyScheduledEventData = ModifyScheduledEventData
+  { modifyScheduledEventDataChannelId    :: Maybe (Maybe ChannelId)
+  , modifyScheduledEventDataLocation     :: Maybe (Maybe T.Text)
+  , modifyScheduledEventDataName         :: Maybe T.Text
+  , modifyScheduledEventDataPrivacyLevel :: Maybe ScheduledEventPrivacyLevel
+  , modifyScheduledEventDataStartTime    :: Maybe UTCTime
+  , modifyScheduledEventDataEndTime      :: Maybe UTCTime
+  , modifyScheduledEventDataDescription  :: Maybe (Maybe T.Text)
+  , modifyScheduledEventDataType         :: Maybe ScheduledEventType
+  , modifyScheduledEventDataStatus       :: Maybe ScheduledEventStatus
+  , modifyScheduledEventDataImage        :: Maybe CreateScheduledEventImage
+  }
+
+instance Default ModifyScheduledEventData where
+  def = ModifyScheduledEventData Nothing
+                                 Nothing
+                                 Nothing
+                                 Nothing
+                                 Nothing
+                                 Nothing
+                                 Nothing
+                                 Nothing
+                                 Nothing
+                                 Nothing
+
+instance ToJSON ModifyScheduledEventData where
+  toJSON ModifyScheduledEventData {..} = object
+    [ (name, value)
+    | (name, Just value) <-
+      [ ("channel_id"          , cid)
+      , ("entity_metadata"     , loc)
+      , ("name", toJSON <$> modifyScheduledEventDataName)
+      , ("scheduled_start_time", toJSON <$> modifyScheduledEventDataStartTime)
+      , ("scheduled_end_time", toJSON <$> modifyScheduledEventDataEndTime)
+      , ("description"         , desc)
+      , ("entity_type", toJSON <$> modifyScheduledEventDataType)
+      , ("status", toJSON <$> modifyScheduledEventDataStatus)
+      , ("image", toJSON <$> modifyScheduledEventDataImage)
+      ]
+    ]
+   where
+    cid = if isNothing modifyScheduledEventDataChannelId
+      then Nothing
+      else
+        let Just cid' = modifyScheduledEventDataChannelId
+        in  if isNothing cid'
+              then Just Null
+              else let Just cid'' = cid' in Just $ toJSON cid''
+    loc = if isNothing modifyScheduledEventDataLocation
+      then Nothing
+      else
+        let Just loc' = modifyScheduledEventDataLocation
+        in  if isNothing loc'
+              then Just Null
+              else
+                let Just loc'' = loc'
+                in  Just $ object [("location", toJSON loc'')]
+    desc = if isNothing modifyScheduledEventDataDescription
+      then Nothing
+      else
+        let Just desc' = modifyScheduledEventDataDescription
+        in  if isNothing desc'
+              then Just Null
+              else let Just desc'' = desc' in Just $ toJSON desc''
```

Comment:
the `toJSON` instance for `ToJSON a => Maybe a` results in `Null` if the maybe is nothing. for the desc and cid at least, this section can be simplified. I'd also recommend that this if else let pattern be changed away from what it is since it's quite verbose and could be just as easily handled by having three cases in a case statement.